### PR TITLE
Support for OpenNebula 4.14

### PR DESCRIPTION
**Pull request by jnoss**
This addresses issue #222 for supporting OpenNebula 4.14

Uses $one_version inside the template for oned.conf, etc, to generate config files for either 4.12 or 4.14 from same template file.

@rwaffen @tuxmea let me know what you think-

### CODE REVIEW

File: templates/oned.conf.erb [hunk 1]
```diff
@@ -28,6 +28,9 @@
 #  scripts.
 #
 #  PORT: Port where oned will listen for xmlrpc calls.
+<% if scope.function_versioncmp([@one_version,'4.14']) >= 0 -%>
+#  LISTEN_ADDRESS: Host IP to listen on for xmlrpc calls (default: all IPs).
+<% end -%>
```
Comment (tuxmea) [inline templates/oned.conf.erb]: This is comment only. I would keep the comment always in place and just add information that this is working with 4.14 and later versions only.

File: templates/oned.conf.erb [hunk 2]
```diff
@@ -72,6 +75,10 @@ SCRIPTS_REMOTE_DIR=/var/tmp/one
 
 PORT = <%= scope.lookupvar('one::oned_port') %>
 
+<% if scope.function_versioncmp([@one_version,'4.14']) >= 0 -%>
```
Comment (tuxmea) [inline templates/oned.conf.erb]: Can you try to keep the variable lookup in the same pattern:
scope.lookupvar('...')
This prevents the dynamic variable lookup from templates which might occur in Puppet 3 and which will stop working in Puppet 4.

File: templates/oned.conf.erb [hunk 3]
```diff
@@ -112,32 +119,47 @@ FEDERATION = [
...
+
 #*******************************************************************************
 # XML-RPC server configuration
 #-------------------------------------------------------------------------------
 #  These are configuration parameters for oned's xmlrpc-c server
 #
 #  MAX_CONN: Maximum number of simultaneous TCP connections the server
-#  will maintain - default 15
+#  will maintain
 #
```
Comment (tuxmea) [inline templates/oned.conf.erb]: Why have the mentioned default values been removed?
I would keep them in the config file.

File: templates/oned.conf.erb [hunk 4]
```diff
@@ -272,10 +294,17 @@ IM_MAD = [
 #    -r number of retries when monitoring a host
 #    -t number of threads, i.e. number of hosts monitored at the same time
 #-------------------------------------------------------------------------------
+<% if scope.function_versioncmp([@one_version,'4.14']) >= 0 -%>
+# IM_MAD = [
+#       name       = "kvm",
+#       executable = "one_im_ssh",
+#       arguments  = "-r 3 -t 15 kvm-probes" ]
```
Comment (tuxmea) [inline templates/oned.conf.erb]: Why disable KVM SSH pull in 4.14? Is this IM_MAD unsupported in 4.14?

File: templates/oned.conf.erb [hunk 5]
```diff
@@ -291,10 +320,17 @@ IM_MAD = [
...
+IM_MAD = [
+    name       = "xen",
+    executable = "one_im_ssh",
+    arguments  = "-r 3 -t 15 xen4" ]
+<% elsif scope.function_versioncmp([@one_version,'4.14']) < 0 -%>
 #IM_MAD = [
 #    name       = "xen",
 #    executable = "one_im_ssh",
 #    arguments  = "-r 3 -t 15 xen4" ]
+<% end -%>
```
Comment (tuxmea) [inline templates/oned.conf.erb]: Why disable Xen in versions prior 4.14?

File: templates/oned.conf.erb [hunk 6]
```diff
@@ -340,6 +376,10 @@ IM_MAD = [
 
 #-------------------------------------------------------------------------------
 #  EC2 Information Driver Manager Configuration
+<% if scope.function_versioncmp([@one_version,'4.14']) >= 0 -%>
+#    -r number of retries when monitoring a host
+#    -t number of threads, i.e. number of hosts monitored at the same time
+<% end -%>
```
Comment (tuxmea) [inline templates/oned.conf.erb]: Same as above: I would keep the documentation inside. Are these parameters only valid in 4.14? In this case just add a comment which versions make use of the parameters.

File: templates/oned.conf.erb [hunk 7]
```diff
@@ -404,6 +444,15 @@ IM_MAD = [
...
 #        An example: "-l migrate=migrate_local,save"
+<% if scope.function_versioncmp([@one_version,'4.14']) >= 0 -%>
+#    -p more than one action per host in parallel, needs support from hypervisor
+#    -s <shell> to execute remote commands, bash by default
+#    -d default snapshot strategy. It can be either 'detach' or 'suspend'. It
+#       defaults to 'suspend'.
+#    -i try to do live-snapshot if available. Currently supported only for kvm
+#       with qcow2 as the datastore TM_MAD. If not available, it will fallback
+#       on the default (-d)
+<% end -%>
```
Comment (tuxmea) [inline templates/oned.conf.erb]: Same here: comment only.

File: templates/oned.conf.erb [hunk 8]
```diff
@@ -432,8 +487,14 @@ VM_MAD = [
 #        overridden for each action.
 #        Valid actions: deploy, shutdown, cancel, save, restore, migrate, poll
 #        An example: "-l migrate,save"
+<% if scope.function_versioncmp([@one_version,'4.14']) >= 0 -%>
+#    -p more than one action per host in parallel, needs support from hypervisor
+#    -s <shell> to execute commands, bash by default
+#    -d default snapshot strategy. It can be either 'detach' or 'suspend'. It
+#       defaults to 'suspend'.
+<% end -%>
```
Comment (tuxmea) [inline templates/oned.conf.erb]: Comment only

File: templates/oned.conf.erb [hunk 9]
```diff
@@ -441,21 +502,36 @@ VM_MAD = [
...
+#
 # Driver for Xen 4.x
+<% if scope.function_versioncmp([@one_version,'4.14']) >= 0 -%>
+VM_MAD = [
+    name       = "xen",
+    executable = "one_vmm_exec",
+    arguments  = "-t 15 -r 0 xen4",
+    default    = "vmm_exec/vmm_exec_xen4.conf",
+    type       = "xen" ]
+<% elsif scope.function_versioncmp([@one_version,'4.14']) < 0 -%>
```
Comment (tuxmea) [inline templates/oned.conf.erb]: enable Xen on 4.14 and later only?

File: templates/oned.conf.erb [hunk 10]
```diff
@@ -441,21 +502,36 @@ VM_MAD = [
...
 #-------------------------------------------------------------------------------
 #  VMware Virtualization Driver Manager Configuration
 #    -r number of retries when monitoring a host
 #    -t number of threads, i.e. number of hosts monitored at the same time
+<% if scope.function_versioncmp([@one_version,'4.14']) >= 0 -%>
+#    -p more than one action per host in parallel, needs support from hypervisor
+#    -s <shell> to execute commands, bash by default
+#    -d default snapshot strategy. It can be either 'detach' or 'suspend'. It
+#       defaults to 'suspend'.
+<% end -%>
```
Comment (tuxmea) [inline templates/oned.conf.erb]: Comment only. Keep it in the config file.

**Comment by jnoss** [comment]
Thanks for taking a look at this. I've updated this PR, the main change is moving the versioncmp out of the templates and using a boolean flag version_gte_4_14 instead (commits 09ffeef and 3c8a96b).

The other changes you mentioned are places where I decided to stick close to the oned.conf from the 4.14 rpm, but, leaving behavior alone in the case of 4.12.  (i.e., it's a no-op except for whitespace for users currently using this puppet module on 4.12).

Also, regarding comments - the sections I separate by version are parameters that are new in 4.14, and I feel it would be difficult to denote a multi-line section as applying to one version but not another.

**Comment by tuxmea** [comment]
I would keep the default values in place (as long as they dont differ between versions.
Otherwise I see no problem with this PR.

**Comment by jnoss** [comment]
The default values don't differ between versions, however I'm trying to stick as close as possible to the oned.conf file from the rpm - which doesn't list the defaults there (neither in 4.12 nor in 4.14). Also, the defaults are present in the params.pp file so users can see them there, and, in the rendered template, the values will be in the section below. If possible, I'd prefer to not hardcode a default value in two different places (the params and the comment).

**Comment by tuxmea** [comment]
@jnoss OK. I agree with avoiding duplications. Good to merge from my point of view.

**Comment by rwaffen** [comment]
thanks for the great work and the comments… only asking my team and than will merge …

**Comment by tuxmea** [comment]
Noone able to merge??

**Comment by nmesstorff** [comment]
@tuxmea i will clear this situation

**Pull request merged**